Confusing output when asserting Iterable<Map<String, ?>> vs DataTable

When executing following test:

```
    @Test
    public void differentTables()
    {
        final TableData tableDataSeparateValues = createTableDataSeparateValues();

        final List<Map<String, ?>> resultData = new LinkedList<>();
        {
            final Map<String, String> row = new HashMap<>();

            row.put("z", "1");
            row.put("y", "2");
            row.put("x", "3");

            resultData.add(row);
        }

        actual(resultData).should(equal(tableDataSeparateValues));

    }
```
output is

```X failed expecting [value] to equal 
:Col A|Col B|Col C:
._____._____._____.
|"v1a"|"v1b"|"v1c"|
._____._____._____|
|"v2a"|"v2b"|"v2c"|
._____._____._____|
: 
    mismatches:
    
    [value]: 
             actual:
             
             :Col A                                    |Col B                                    |Col C                                    :
             ._________________________________________._________________________________________._________________________________________.
             |"***                                     |"***                                     |"***                                     |
             |mismatches:                              |mismatches:                              |mismatches:                              |
             |                                         |                                         |                                         |
             |Col A:   actual: null                    |Col B:   actual: null                    |Col C:   actual: null                    |
             |       expected: "v1a" <java.lang.String>|       expected: "v1b" <java.lang.String>|       expected: "v1c" <java.lang.String>|
             |***                                      |***                                      |***                                      |
             |                                         |                                         |                                         |
             |null"                                    |null"                                    |null"                                    |
             ._________________________________________._________________________________________._________________________________________|
             
             expected:
             
             :Col A                                    |Col B                                    |Col C                                    :
             ._________________________________________._________________________________________._________________________________________.
             |"***                                     |"***                                     |"***                                     |
             |mismatches:                              |mismatches:                              |mismatches:                              |
             |                                         |                                         |                                         |
             |Col A:   actual: null                    |Col B:   actual: null                    |Col C:   actual: null                    |
             |       expected: "v1a" <java.lang.String>|       expected: "v1b" <java.lang.String>|       expected: "v1c" <java.lang.String>|
             |***                                      |***                                      |***                                      |
             |                                         |                                         |                                         |
             |v1a"                                     |v1b"                                     |v1c"                                     |
             ._________________________________________._________________________________________._________________________________________|
             |"v2a"                                    |"v2b"                                    |"v2c"                                    |
             ._________________________________________._________________________________________._________________________________________|
             
             missing rows:
             :Col A|Col B|Col C:
             ._____._____._____.
             |"v2a"|"v2b"|"v2c"|
             ._____._____._____| (144ms)

java.lang.AssertionError: 
mismatches:

[value]: 
         actual:
         
         :Col A                                    |Col B                                    |Col C                                    :
         ._________________________________________._________________________________________._________________________________________.
         |"***                                     |"***                                     |"***                                     |
         |mismatches:                              |mismatches:                              |mismatches:                              |
         |                                         |                                         |                                         |
         |Col A:   actual: null                    |Col B:   actual: null                    |Col C:   actual: null                    |
         |       expected: "v1a" <java.lang.String>|       expected: "v1b" <java.lang.String>|       expected: "v1c" <java.lang.String>|
         |***                                      |***                                      |***                                      |
         |                                         |                                         |                                         |
         |null"                                    |null"                                    |null"                                    |
         ._________________________________________._________________________________________._________________________________________|
         
         expected:
         
         :Col A                                    |Col B                                    |Col C                                    :
         ._________________________________________._________________________________________._________________________________________.
         |"***                                     |"***                                     |"***                                     |
         |mismatches:                              |mismatches:                              |mismatches:                              |
         |                                         |                                         |                                         |
         |Col A:   actual: null                    |Col B:   actual: null                    |Col C:   actual: null                    |
         |       expected: "v1a" <java.lang.String>|       expected: "v1b" <java.lang.String>|       expected: "v1c" <java.lang.String>|
         |***                                      |***                                      |***                                      |
         |                                         |                                         |                                         |
         |v1a"                                     |v1b"                                     |v1c"                                     |
         ._________________________________________._________________________________________._________________________________________|
         |"v2a"                                    |"v2b"                                    |"v2c"                                    |
         ._________________________________________._________________________________________._________________________________________|
         
         missing rows:
         :Col A|Col B|Col C:
         ._____._____._____.
         |"v2a"|"v2b"|"v2c"|
         ._____._____._____|

	at ... 


Process finished with exit code 255
```

where expected the output to match table to table comparison:

```
    @Test
    public void differentTables()
    {
        final TableData tableDataSeparateValues = createTableDataSeparateValues();

        final List<Map<String, ?>> resultData = new LinkedList<>();
        {
            final Map<String, String> row = new HashMap<>();

            row.put("z", "1");
            row.put("y", "2");
            row.put("x", "3");

            resultData.add(row);
        }

        actual(resultData).should(equal(tableDataSeparateValues));

    }
```


with output

```
X failed expecting [value] to equal 
:Col A|Col B|Col C:
._____._____._____.
|"v1a"|"v1b"|"v1c"|
._____._____._____|
|"v2a"|"v2b"|"v2c"|
._____._____._____|
: 
    mismatches:
    
    [value]: missing columns: Col A, Col B, Col C
             actual:
             
             :z  |y|x:
             .___._._.
             |"1"|2|3|
             .___._._|
             
             expected:
             
             :Col A|Col B|Col C:
             ._____._____._____.
             |"v1a"|"v1b"|"v1c"|
             ._____._____._____|
             |"v2a"|"v2b"|"v2c"|
             ._____._____._____|
             
             missing rows:
             :Col A|Col B|Col C:
             ._____._____._____.
             |"v2a"|"v2b"|"v2c"|
             ._____._____._____| (24ms)

java.lang.AssertionError: 
mismatches:

[value]: missing columns: Col A, Col B, Col C
         actual:
         
         :z  |y|x:
         .___._._.
         |"1"|2|3|
         .___._._|
         
         expected:
         
         :Col A|Col B|Col C:
         ._____._____._____.
         |"v1a"|"v1b"|"v1c"|
         ._____._____._____|
         |"v2a"|"v2b"|"v2c"|
         ._____._____._____|
         
         missing rows:
         :Col A|Col B|Col C:
         ._____._____._____.
         |"v2a"|"v2b"|"v2c"|
         ._____._____._____|

	at org.testingisdocumenting.webtau.expectation.ActualValue.handleMismatch(ActualValue.java:143)
```

Auto Conversion from Map seems to use columns from expected table rather than columns that are present in the map (as map keys)

--- a/webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/equality/handlers/IterableAndTableDataCompareToHandler.java
+++ b/webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/equality/handlers/IterableAndTableDataCompareToHandler.java
@@ -27,9 +27,8 @@ import org.testingisdocumenting.webtau.expectation.ActualPath;
 import org.testingisdocumenting.webtau.expectation.equality.CompareToComparator;
 import org.testingisdocumenting.webtau.expectation.equality.CompareToHandler;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
+import java.util.stream.Stream;
 
 public class IterableAndTableDataCompareToHandler implements CompareToHandler {
     @Override
@@ -41,22 +40,47 @@ public class IterableAndTableDataCompareToHandler implements CompareToHandler {
     @SuppressWarnings("unchecked")
     public void compareEqualOnly(CompareToComparator comparator, ActualPath actualPath, Object actual, Object expected) {
         TableData expectedTable = (TableData) expected;
-        TableData actualTable = createTableFromIterable(expectedTable.getHeader(), (Iterable<Object>) actual);
+        TableData actualTable = createTableFromIterable((Iterable<Object>) actual);
 
         TableDataComparisonResult result = TableDataComparison.compare(actualTable, expectedTable);
-        if (! result.areEqual()) {
+        if (!result.areEqual()) {
             comparator.reportNotEqual(this, actualPath, new TableDataComparisonReport(result).generate());
         }
     }
 
-    private static TableData createTableFromIterable(TableDataHeader expectedHeader, Iterable<Object> actualList) {
-        TableData actualTable = new TableData(expectedHeader.getNamesStream());
-        for (Object actualRecord : actualList) {
-            Map<String, ?> actualMap = ToMapConverters.convert(actualRecord);
-            actualTable.addRow(mapToList(expectedHeader, actualMap));
+    private static TableData createTableFromIterable(Iterable<Object> actualList) {
+        Iterator<Object> actualIt = actualList.iterator();
+        if (!actualIt.hasNext()) {
+            return new TableData(Stream.empty());
+        }
+
+        List<Map<String, ?>> actualAsListOfMaps = convertActualToListOfMaps(actualList);
+        TableDataHeader actualHeader = createCombinedActualHeader(actualAsListOfMaps);
+
+        TableData actualTable = new TableData(actualHeader);
+        for (Map<String, ?> actualRow : actualAsListOfMaps) {
+            actualTable.addRow(mapToList(actualHeader, actualRow));
         }
 
         return actualTable;
+    }
+
+    private static List<Map<String, ?>> convertActualToListOfMaps(Iterable<Object> actualIt) {
+        List<Map<String, ?>> result = new ArrayList<>();
+        for (Object row : actualIt) {
+            result.add(ToMapConverters.convert(row));
+        }
+
+        return result;
+    }
+
+    private static TableDataHeader createCombinedActualHeader(List<Map<String, ?>> actual) {
+        Set<String> columnNames = new LinkedHashSet<>();
+        for (Map<String, ?> row : actual) {
+            columnNames.addAll(row.keySet());
+        }
+
+        return new TableDataHeader(columnNames.stream());
     }
 
     private static List<Object> mapToList(TableDataHeader header, Map<String, ?> map) {

--- a/webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/equality/handlers/TableDataCompareToHandler.java
+++ b/webtau-core/src/main/java/org/testingisdocumenting/webtau/expectation/equality/handlers/TableDataCompareToHandler.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2022 webtau maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -33,7 +34,7 @@ public class TableDataCompareToHandler implements CompareToHandler {
     @Override
     public void compareEqualOnly(CompareToComparator comparator, ActualPath actualPath, Object actual, Object expected) {
         TableDataComparisonResult result = TableDataComparison.compare((TableData) actual, (TableData) expected);
-        if (! result.areEqual()) {
+        if (!result.areEqual()) {
             comparator.reportNotEqual(this, actualPath, new TableDataComparisonReport(result).generate());
         }
     }

--- a/webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/equality/handlers/IterableAndTableDataCompareToHandlerTest.groovy
+++ b/webtau-core/src/test/groovy/org/testingisdocumenting/webtau/expectation/equality/handlers/IterableAndTableDataCompareToHandlerTest.groovy
@@ -19,6 +19,7 @@ package org.testingisdocumenting.webtau.expectation.equality.handlers
 
 import groovy.transform.Canonical
 import groovy.transform.Sortable
+import org.testingisdocumenting.webtau.data.table.TableData
 import org.testingisdocumenting.webtau.expectation.equality.CompareToComparator
 import org.junit.Test
 
@@ -69,6 +70,24 @@ class IterableAndTableDataCompareToHandlerTest {
 
         def report = comparator.generateEqualMismatchReport()
         assert report.contains("lot:   actual: 1.0")
+    }
+
+    @Test
+    void "should report columns mismatch"() {
+        def listOfMaps = [
+                [a: 1, b: 2, C: 3]
+        ]
+
+        TableData table = table("A", "B", "C",
+                               ___________________,
+                                 1,   2,  2)
+
+        def comparator = CompareToComparator.comparator()
+        assert !comparator.compareIsEqual(createActualPath("maps"), listOfMaps, table)
+
+        def report = comparator.generateEqualMismatchReport()
+        assert report.contains("missing columns: A, B")
+        assert report.contains("C:   actual: 3")
     }
 
     @Canonical

--- a/webtau-docs/znai/release-notes/1.50/add-2022-11-07-list-of-maps-table-fix.md
+++ b/webtau-docs/znai/release-notes/1.50/add-2022-11-07-list-of-maps-table-fix.md
@@ -1,0 +1,1 @@
+* Fix: properly report missing columns in case of list of maps and [TableData](reference/table-data) comparison

